lib/remote_storage/couchdb.rb

```diff
@@ -0,0 +1,458 @@
+
+    def parent_directories_for(directory)
+      directories = directory.split("/")
+      parent_directories = []
+
+      while directories.any?
+        parent_directories << directories.join("/")
+        directories.pop
+      end
+
+      parent_directories << "" # add empty string for the root directory
+
+      parent_directories
+    end
+
+    def top_directory(directory)
+      if directory.match(/\//)
+        directory.split("/").last
+      elsif directory != ""
+        return directory
+      end
+    end
+
+    def parent_directory_for(directory)
+      if directory.match(/\//)
+        return directory[0..directory.rindex("/")]
+      elsif directory != ""
+        return "/"
+      end
+    end
+
+    def update_metadata_object(user, directory, key, metadata)
+      redis_key = redis_metadata_object_key(user, directory, key)
+      redis.hmset(redis_key, *metadata)
+      redis.sadd "rs:m:#{user}:#{directory}/:items", key
+
+      true
+    end
+
+    def update_dir_objects(user, directory, timestamp, checksum)
+      parent_directories_for(directory).each do |dir|
+        etag = etag_for(dir, timestamp, checksum)
+
+        key = "rs:m:#{user}:#{dir}/"
+        metadata = {e: etag, m: timestamp}
+        redis.hmset(key, *metadata)
+        redis.sadd "rs:m:#{user}:#{parent_directory_for(dir)}:items", "#{top_directory(dir)}/"
+      end
+    end
+
+    def delete_metadata_objects(user, directory, key)
+      redis.del redis_metadata_object_key(user, directory, key)
+      redis.srem "rs:m:#{user}:#{directory}/:items", key
+    end
+
+    def delete_dir_objects(user, directory)
+      timestamp = (Time.now.to_f * 1000).to_i
+
+      parent_directories_for(directory).each do |dir|
+        if dir_empty?(user, dir)
+          redis.del "rs:m:#{user}:#{dir}/"
+          redis.srem "rs:m:#{user}:#{parent_directory_for(dir)}:items", "#{top_directory(dir)}/"
+        else
+          etag = etag_for(dir, timestamp)
+
+          metadata = {e: etag, m: timestamp}
+          redis.hmset("rs:m:#{user}:#{dir}/", *metadata)
+        end
+      end
+    end
+
+    def dir_empty?(user, dir)
+      redis.smembers("rs:m:#{user}:#{dir}/:items").empty?
+    end
+
+    def redis_metadata_object_key(user, directory, key)
+      "rs:m:#{user}:#{[directory, key].delete_if(&:empty?).join("/")}"
+    end
+
+    def url_for_key(user, directory, key)
+      File.join [base_url, escape(File.join([user, directory, key].reject(&:empty?)))].compact
+    end
+
+    def base_url
+      @base_url ||= settings.couchdb["uri"]
+    end
+
+    def default_headers
+      {}
+    end
+
+    def do_put_request(url, data, content_type)
+      begin
+        res = RestClient.get(url)
+        rev = JSON.parse(res.body)["_rev"]
+        url = "#{url}?rev=#{rev}"
+      rescue RestClient::ResourceNotFound
+        #do nothing
+      end
+      mime_type = MIME::Types[content_type].first
+      if mime_type.content_type == "application/json" || !mime_type.binary?
```

Comment:
When `mime_type.content_type` is "application/json", then `mime_type.binary?` should also be `false`, no?

Then `if !mime_type.binary?` should be enought here.